secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/trusteddirectories/FetchTrustedDirectoryFilter.java

```diff
@@ -0,0 +1,127 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.sapi.gateway.trusteddirectories;
+
+import static org.forgerock.openig.util.JsonValues.requiredHeapObject;
+
+import org.forgerock.http.Filter;
+import org.forgerock.http.Handler;
+import org.forgerock.http.protocol.Request;
+import org.forgerock.http.protocol.Response;
+import org.forgerock.json.jose.jwt.JwtClaimsSet;
+import org.forgerock.openig.heap.GenericHeaplet;
+import org.forgerock.openig.heap.HeapException;
+import org.forgerock.services.context.AttributesContext;
+import org.forgerock.services.context.Context;
+import org.forgerock.util.Reject;
+import org.forgerock.util.promise.NeverThrowsException;
+import org.forgerock.util.promise.Promise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.forgerock.sapi.gateway.dcr.ApiClient;
+import com.forgerock.sapi.gateway.dcr.FetchApiClientFilter;
+import com.forgerock.sapi.gateway.fapi.FAPIUtils;
+
+/**
+ * Fetches {@link TrustedDirectory} configuration for the {@link ApiClient} that is configured in the {@link AttributesContext}
+ * and adds the TrustedDirectory to the attributes context so that it can be used by subsequent filters.
+ *
+ * This filter must be installed after a filter which adds the ApiClient to the context, typically: {@link FetchApiClientFilter}
+ */
+public class FetchTrustedDirectoryFilter implements Filter {
+
+    /**
+     * The key to use to get the TrustedDirectory from the AttributesContext
+     */
+    public static final String TRUSTED_DIRECTORY_ATTR_KEY = "trustedDirectory";
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * The Service used to retrieve the TrustedDirectory
+     */
+    private final TrustedDirectoryService trustedDirectoryService;
+
+    /**
+     * Utility method to retrieve a TrustedDirectory object from a Context.
+     * This method can be used by other filters to retrieve the TrustedDirectory installed into the attributes context by
+     * this filter.
+     *
+     * @param context the context to retrieve the TrustedDirectory from
+     * @return the TrustedDirectory or null if it is not set in the context.
+     */
+    public static TrustedDirectory getTrustedDirectoryFromContext(Context context) {
+        return (TrustedDirectory) context.asContext(AttributesContext.class).getAttributes().get(TRUSTED_DIRECTORY_ATTR_KEY);
+    }
+
+    public FetchTrustedDirectoryFilter(TrustedDirectoryService trustedDirectoryService) {
+        Reject.ifNull(trustedDirectoryService, "trustedDirectoryService must be provided");
+        this.trustedDirectoryService = trustedDirectoryService;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
```

Comment:
Ah, what's that NeverThrowsException do?? Maybe it answers all my questions? Still some logging and nice error messages still nice to have

Reply:
It is a marker to show that a methods never throws an exception. 

All of the HTTP based methods will have this sig, it encourages you to return a HTTP based error message.

Runtime exceptions may still be thrown, the `org.forgerock.openig.filter.RuntimeExceptionFilter` acts as an uncaught exception handler in the chain and will catch RuntimeExceptions and return a HTTP 500 response (if they are not caught and handled by another filter in the chain).